src/engine/source/_builder/asset.hpp

```diff
@@ -0,0 +1,128 @@
+#ifndef _ASSET_H
+#define _ASSET_H
+
+#include <string>
+#include <unordered_set>
+
+#include <fmt/format.h>
+
+#include "_builder/expression.hpp"
+#include "_builder/json.hpp"
+#include "_builder/registry.hpp"
+
+class Asset
+{
+public:
+    enum class Type
+    {
+        DECODER,
+        FILTER,
+        RULE,
+        OUTPUT
+    };
+
+    std::string m_name;
+    Expression m_check;
+    Expression m_stages;
+    Type m_type;
+    std::unordered_set<std::string> m_parents;
+    Json m_metadata;
+
+    Asset(std::string name, Type type)
+        : m_name {name}
+        , m_type {type}
+    {
+    }
+
+    Asset(const Json& jsonDefinition, Type type)
+        : m_type {type}
+    {
+        auto objectDefinition = jsonDefinition.getObject();
+
+        // Get name
+        auto namePos = std::find_if(objectDefinition.begin(),
+                                    objectDefinition.end(),
+                                    [](auto tuple)
+                                    { return std::get<0>(tuple) == "name"; });
+        if (namePos != objectDefinition.end())
+        {
+            m_name = std::get<1>(*namePos).getString();
+            objectDefinition.erase(namePos);
+        }
+        else
+        {
+            throw std::runtime_error("Asset definition must have a name");
+        }
+
```

Comment:
This block gets repeated several times, can we assign an enum on each stage (with it's weight) and put it inside a loop.
First Stage::name = 0, Stage::parent and so on.
This way we also avoid those raw strings and in the future, we could add stages without modifying all the methods.
This could also imply modifying how we are treating m_chack and m_stages Expressions.